### PR TITLE
Tests to validate resources deployed by etcd-druid

### PR DESCRIPTION
**Pull request by georgekuruvillak**
**What this PR does / why we need it**:
This PR consists of test to validate the functionality of tests.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: controllers/etcd_controller_test.go [hunk 1]
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
...
+			Expect(instance.Spec.Etcd.ClientPort).To(BeNil())
+			Expect(port.ContainerPort).To(Equal(clientPort))
+		}
+	}
+	Expect(instance.Spec.Etcd.Image).To(BeNil())
+	imageVector, err := imagevector.ReadGlobalImageVectorWithEnvOverride(getImageYAMLPath())
+	Expect(err).NotTo(HaveOccurred())
+	images, err := imagevector.FindImages(imageVector, imageNames)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(containers[common.Etcd].Image).To(Equal(fmt.Sprintf("%s:%s", images[common.Etcd].Repository, *images[common.Etcd].Tag)))
```
Comment (amshuman-kr) [inline controllers/etcd_controller_test.go:386]: https://onsi.github.io/gomega/#consistofelement-interface?
https://onsi.github.io/gomega/#containelementelement-interface?

File: controllers/etcd_controller_test.go [hunk 2]
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
...
+
+	// Validate Metrics MetricsLevel
+	Expect(instance.Spec.Etcd.Metrics).To(BeNil())
+	val, ok = config[metricsKey]
```
Comment (amshuman-kr) [inline controllers/etcd_controller_test.go:390]: https://onsi.github.io/gomega/#havekeywithvaluekey-interface-value-interface

File: controllers/etcd_controller_test.go [hunk 3]
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
 
 }
 
+func validateEtcdConfigWithDefaults(s *appsv1.StatefulSet, cm *corev1.ConfigMap, svc *corev1.Service, instance *druidv1alpha1.Etcd) {
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply controllers/etcd_controller_test.go:340]: I have removed all the loops and conditionals and replaced them with gomega structs. The tests are much cleaner now. PTAL

File: controllers/etcd_controller_test.go [hunk 4]
```diff
@@ -271,6 +340,367 @@ func podDeleted(c client.Client, etcd *druidv1alpha1.Etcd) error {
...
+	Expect(instance.Spec.Etcd.TLS).To(BeNil())
+	// the scheme in readiness probe should be https
+	readinessProbe := containers[common.Etcd].ReadinessProbe
+	Expect(readinessProbe.HTTPGet.Scheme).To(Equal(corev1.URISchemeHTTP))
+	// command slice should have only 7 entries. TLS specific fields would increase the count to more than 7.
+	livelinessProbe := containers[common.Etcd].LivenessProbe
+	Expect(len(livelinessProbe.Exec.Command)).To(Equal(7))
+}
+
+func validateEtcdConfig(s *appsv1.StatefulSet, cm *corev1.ConfigMap, svc *corev1.Service, instance *druidv1alpha1.Etcd) {
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply controllers/etcd_controller_test.go:417]: I have removed all the loops and conditionals and replaced them with gomega structs. The tests are much cleaner now. PTAL

File: controllers/etcd_controller_test.go (hunk 1 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller_test.go:385]: It is just a string comparison here. Equal should suffice, ryt?

**Comment by georgekuruvillak** [comment]
@AndreasBurger Can you check out the pipeline_definition file in `.ci` directory.

**Comment by amshuman-kr** [comment]
@georgekuruvillak I don't see any changes related to gomega structs.

**Comment by georgekuruvillak** [comment]
@amshuman-kr I haven't added gstruct. Instead, I have removed all the loops and conditionals and replaced them with gomega structs. The tests look much cleaner now. I shall add in gostruct too.

**Comment by amshuman-kr** [comment]
@georgekuruvillak I didn't understand what you meant by gomega structs then? I still see loops and conditions.

**Comment by georgekuruvillak** [comment]
The `if` condition is used [here]( https://github.com/gardener/etcd-druid/pull/41/files#diff-03838f4ddc16876e7feda81a4bedf7d8R467) as the `tls` fields are used to check the liveliness probes and readiness probes. The `for` loop is used [here](https://github.com/gardener/etcd-druid/pull/41/files#diff-03838f4ddc16876e7feda81a4bedf7d8R427) for choosing the container to test against. By go omega struct I meant the constructs in gomega like `ConsistOf`, `HasKeyWithValue`. I havent implemented gstruct.

**Comment by swapnilgm** [comment]
@georgekuruvillak Please remove pipeline definitions updates. Those are fixed with https://github.com/gardener/etcd-druid/pull/58

**Comment by georgekuruvillak** [comment]
@amshuman-kr I have made the changes. PTAL.

File: controllers/constants_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            echo "Requesting sidecar to perform full validation"
+            check_and_start_etcd full
+      fi
+fi`
+)
+
+var (
+	bootstrapScriptWithoutTLS = scriptWithoutTLS + startManagedEtcd + checkAndStartEtcdWithoutTLS + bootstrapWithValidation
+	bootstrapScriptWithTLS    = scriptWithTLS + startManagedEtcd + checkAndStartEtcdWithTLS + bootstrapWithValidation
+	cmdIterator               = func(element interface{}) string {
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply controllers/constants_test.go:136]: Makes sense. Ill make sthe change and update the PR.

**Comment by swapnilgm** [comment]
@georgekuruvillak Will you please squash the commits.

**Pull request merged**